src/main/java/com/epam/aidial/core/service/CacheService.java

```diff
@@ -0,0 +1,447 @@
+
+        if (!cacheRecord.exists()) {
+            return Result.empty();
+        }
+
+        if (cacheRecord.binary) {
+            throw new IllegalStateException("The value at %s is not an array".formatted(key));
+        }
+
+        return new Result<>(cacheRecord.item.withBody(BASE64_DECODER.decode(cacheRecord.item.body)));
+    }
+
+    @Nullable
+    private Record redisGet(String key, boolean withBody) {
+        RMap<String, String> map = redis.getMap(key, StringCodec.INSTANCE);
+        Map<String, String> fields = map.getAll(withBody ? REDIS_FIELDS : REDIS_FIELDS_NO_BODY);
+
+        if (fields.isEmpty()) {
+            return null;
+        }
+
+        boolean exists = Boolean.parseBoolean(Objects.requireNonNull(fields.get(EXISTS_ATTRIBUTE)));
+        boolean binary = Boolean.parseBoolean(Objects.requireNonNull(fields.get(BINARY_ATTRIBUTE)));
+        boolean synced = Boolean.parseBoolean(Objects.requireNonNull(fields.get(SYNCED_ATTRIBUTE)));
+        if (!exists) {
+            return new Record(null, binary, synced);
+        }
+
+        String body = fields.get(BODY_ATTRIBUTE);
+        String etag = ResourceUtil.extractEtag(fields);
+        String contentType = StringUtils.defaultIfEmpty(fields.get(CONTENT_TYPE_ATTRIBUTE), null);
+        Long contentLength = nullIfEmpty(fields.get(CONTENT_LENGTH_ATTRIBUTE));
+        Long createdAt = nullIfEmpty(fields.get(ResourceUtil.CREATED_AT_ATTRIBUTE));
+        Long updatedAt = nullIfEmpty(fields.get(ResourceUtil.UPDATED_AT_ATTRIBUTE));
+
+        return new Record(
+                new Item<>(new ItemMetadata(etag, createdAt, updatedAt, contentType, contentLength), body),
+                binary,
+                synced);
+    }
+
+    private void redisPut(String key, Record cacheRecord) {
+        RScoredSortedSet<String> set = redis.getScoredSortedSet(resourceQueue, StringCodec.INSTANCE);
+        set.add(time() + syncDelay, key); // add resource to sync set before changing because calls below can fail
+
+        RMap<String, String> map = redis.getMap(key, StringCodec.INSTANCE);
+
+        if (!cacheRecord.synced) {
+            map.clearExpire();
+        }
+
+        Map<String, String> fields = new HashMap<>();
+        if (cacheRecord.exists()) {
+            Item<String> item = cacheRecord.item;
+            fields.put(BODY_ATTRIBUTE, item.body);
+            ItemMetadata metadata = item.metadata;
+            fields.put(ResourceUtil.ETAG_ATTRIBUTE, metadata.etag);
+            fields.put(ResourceUtil.CREATED_AT_ATTRIBUTE, emptyIfNull(metadata.createdAt));
+            fields.put(ResourceUtil.UPDATED_AT_ATTRIBUTE, emptyIfNull(metadata.updatedAt));
+            fields.put(CONTENT_TYPE_ATTRIBUTE, emptyIfNull(metadata.contentType));
+            fields.put(CONTENT_LENGTH_ATTRIBUTE, emptyIfNull(metadata.contentLength));
+            fields.put(EXISTS_ATTRIBUTE, String.valueOf(true));
+        } else {
+            REDIS_FIELDS.forEach(field -> fields.put(field, ""));
```

Comment:
Will it fail on redisGet?

Reply:
In redisGet there is a check for "exists" flag.